File: PRACTICAS/practica1/Civitas_Ruby/lib/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
class TituloPropiedad
  
    @@FACTORINTERESESHIPOTECA = 1.1
 
  
  def initialize(nombre,prec_alq_base,fact_rev,prec_baship,prec_compr,prec_edif)
    @nombre = nombre
    @alquiler_base = prec_alq_base
    @factor_revalorizacion = fact_rev
    @precio_hipoteca_base = prec_baship
    @precio_compra = prec_compr
    @precio_edificar = prec_edif
    @propietario = null
    @hipotecado = false
    @num_casas = @num_hoteles = 0
    
  end
  
  def actualiza_propietario_por_conversion(jugador)
   
    @propietario = jugador
    
  end
  
  def cancelar_hipoteca(jugador)
    
    if(!@hipotecado && es_este_el_propietario(jugador))
            
      jugador.paga(get_importe_cancelar_hipoteca);
      @hipotecado = false;
      return true;
            
    else
      return false 
        
    
    end
  end
  
  
  def cantidad_casas_hoteles()
    
    cant = get_num_casas + get_num_hoteles
    return cant
    
  end
  
  
  def comprar(jugador)
  
    if(tiene_propietario)
            
      return false
        
    else
            
      actualiza_propietario_por_conversion(jugador);
      jugador.paga(get_precio_compra);
      return true    
            
    end
  end
  
  def construir_casa(jugador)
   
    prop = false 
    if(es_este_el_propietario(jugador))
            
      jugador.paga(get_precio_compra);
      @num_casas+=1
      prop = true
      return prop
    end
        return prop
    
  end
  
  def construir_hotel(jugador)
    
    prop = false 
    if(es_este_el_propietario(jugador))
            
      jugador.paga(get_precio_compra)
      @num_hoteles+=1
      prop = true
      return prop
    end
        return prop
    
  end
  
  def derruir_casas(n,jugador)
    
    if(es_este_el_propietario(jugador) && get_num_casas>=n)
      @num_casas = @num_casas-n
      return true
            
    else
      return false
    
    end
  end
  
  
  def es_este_el_propietario(jugador)
    
    return @propietario == jugador
    
  end
  
  
  private :es_este_el_propietario
  
  attr_reader :hipotecado    #ESTE ES PUBLIC PERO NO SE COMO SE PONE
  
  
  def get_importe_cancelar_hipoteca()
    
    
    cantidad_recibida = (@precio_hipoteca_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5)));
    return cantidad_recibida*@@FACTORINTERESESHIPOTECA;
    
  end
  
  
  attr_reader :precio_hipoteca_base
  
  private :precio_hipoteca_base  #creo que esto ya esta puesto en la declaracion del atributo
  
  
  attr_reader :nombre, :num_casas, :num_hoteles
  
  
  def get_precio_alquiler()
 
        precio_alquiler = (@precio_alquiler_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5)));
        
        if (@hipotecado || propietario_encarcelado)
            
          precio_alquiler=0;
         
        end
        return precio_alquiler;
  end
    
        
  
  private :get_precio_alquiler
  
  attr_reader :precio_compra, :precio_edificar
  
  def get_precio_venta()
    
    sum = get_precio_compra + (get_precio_edificar * @factor_revalorizacion)
        return sum 
  end
  
  private :get_precio_venta
  
  attr_reader :propietario
  
  def hipotecar(jugador)
    
    if (!@hipotecado && es_este_el_propietario(jugador))
            
      jugador.recibe(get_importe_hipoteca) 
      @hipotecado = true
      return true
    else
            
      return false
            
    end
    
  end
  
  def propietario_encarcelado()
    
    if(propietario.is_encarcelado)
      return true
    else
      return false
    
    end
  end
  
  private :propietario_encarcelado
  
  def tiene_propietario()
    
    return @propietario!=null
    
  end
  
  
  def to_string()
    
    if (@hipotecado)
      return "La propiedad #{@nombre} pertenece a #{@propietario.nombre} y está hipotecada. Tiene #{@num_casas} casas, #{@num_hoteles}
              hoteles, precio de compra #{@precio_compra} $, precio edificar #{@precio_edificar} $, factor de revalorizacion #{@factor_revalorizacion}
              , alquiler base de #{@alquiler_base} $ y una hipoteca base de #{@precio_hipoteca_base} $.";
    else
      return "La propiedad #{@nombre} pertenece a #{@propietario.nombre} y no está hipotecada. Tiene #{@num_casas} casas, #{@num_hoteles}
              hoteles, precio de compra #{@precio_compra} $, precio edificar #{@precio_edificar} $, factor de revalorizacion #{@factor_revalorizacion}
              , alquiler base de #{@alquiler_base} $ y una hipoteca base de #{@precio_hipoteca_base} $.";
    
    end
  end
  
  public :to_string
  
  def tramitar_alquiler(jugador)
    
    if(tiene_propietario && !es_este_el_propietario(jugador))
            
      jugador.paga_alquiler(get_precio_alquiler)
      @propietario.recibe(get_precio_alquiler)  #este propietario es del attr_reader
            
    end
    
  end
  
  
  def vender(jugador)
    
    if(es_este_el_propietario(jugador)&& !@hipotecado)
            
      jugador.recibe(get_precio_venta)
      @propietario=null
      derruir_casas(@num_casas, jugador) #este numCasas es del attr_reader
      @num_hoteles=0
      return true
            
    else
      return false
    
    end
  end
  
  
end
end

File: PRACTICAS/CAMBIAR_RUBY/COMPROBACIONES JAVA/src/civitas_p1/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

/**
 *
 * @author angel
 */
public class TituloPropiedad {
    
    private String nombre;
    private float alquilerBase;
    private float hipotecaBase;
    private float precioCompra;
    private float precioEdificar;
    private Jugador propietario;
    private static final float FACTORINTERESHIPOTECA = (float) 1.1;
    private float factorRevalorizacion;
    private boolean hipotecado;
    private int numCasas;
    public int numHoteles;
            
    
    TituloPropiedad(String name, float pBaseAlquiler, float fRevalorizacion, float pBaseHipoteca, float pCompra, float pEdificar  ){
        
        nombre = name;
        alquilerBase = pBaseAlquiler;
        hipotecaBase = pBaseHipoteca;
        precioCompra = pCompra;      
        precioEdificar = pEdificar;
        propietario = null;
        factorRevalorizacion = fRevalorizacion;
        hipotecado = false;
        numCasas = 0;
        numHoteles = 0;        
        
    }
    
    @Override
    public String toString(){
        
        if (hipotecado)
            return "La propiedad " + nombre + " pertenece a " + propietario.getNombre() + " y está hipotecada. Tiene " + numCasas + " casas, "
                    + numHoteles + " hoteles, precio de compra " + precioCompra +" $, precio edificar " + precioEdificar + " $, factor de revalorizacion = "
                    + factorRevalorizacion + ", alquiler base de " + alquilerBase + " $ y una hipoteca base de " + hipotecaBase + " $.";
        else
            return "La propiedad " + nombre + " pertenece a " + propietario.getNombre() + " y no está hipotecada. Tiene " + numCasas + " casas, "
                    + numHoteles + " hoteles, precio de compra " + precioCompra +" $, precio edificar " + precioEdificar + " $, factor de revalorizacion = "
                    + factorRevalorizacion + ", alquiler base de " + alquilerBase + " $ y una hipoteca base de " + hipotecaBase + " $.";
        
    }
    
    
    
    public float getPrecioAlquiler(){
        if (hipotecado || propietario.isEncarcelado())
            return 0;
        else
            return (float) (alquilerBase*(1+(numCasas*0.5)+(numHoteles * 2.5)));
        
    }
    
    float getImporteCancelarHipoteca(){
    
        return hipotecaBase * FACTORINTERESHIPOTECA;
    
    }   
    
    boolean cancelarHipoteca(Jugador jugador){
        
        boolean result = false; 
        
        if(hipotecado && esEsteElPropietario(jugador)){      
            jugador.paga(getImporteCancelarHipoteca());
            result = true;
            hipotecado = false;
        }
        
        return result;
        
    }
    
    String getNombre(){
        return nombre;
    }
    
    Boolean hipotecar(Jugador jugador){
        
        if (!hipotecado && esEsteElPropietario(jugador)){
            hipotecado = true;
            jugador.modificarSaldo(hipotecaBase);
            return true;
        }
        else
            return false;
        
    }
    
    void tramitarAlquiler (Jugador jugador){
        
        float precio;
        
        if(tienePropietario()){
        
            if (!esEsteElPropietario(jugador)){
            precio = getPrecioAlquiler();
            jugador.pagaAlquiler(precio);
            propietario.recibe(precio);
                
            }
            
            
            
        }
                  
    }
    
    private boolean propietarioEncarcelado(){
        
        if (propietario != null && propietario.isEncarcelado()) 
            return true;
        else
            return false;
        
    }
    
    int cantidadCasasHoteles(){
        return numCasas + numHoteles;
    }
    
    boolean derruirCasas( int n, Jugador jugador){
        
        if (esEsteElPropietario(jugador) && n <= numCasas){ //esEsteElPropietario(Jugador jugador)
            numCasas -= n;
            return true;
            
        }
        else
            return false;
    }
    
    private float getPrecioVenta(){
        return precioCompra + (numCasas + 5*numHoteles) * precioEdificar * factorRevalorizacion;
    }
    
    
    boolean construirCasa(Jugador jugador){
        
        boolean construir = false;
        
        if (esEsteElPropietario(jugador)){ 
            jugador.modificarSaldo(precioEdificar*(-1));
            numCasas++;
            construir = true;
        }
        
        return construir;
        
    }
    
    boolean construirHotel(Jugador jugador){
        
        boolean result = false;
        
        if (esEsteElPropietario(jugador)){ 
            propietario.paga(precioEdificar); 
            numHoteles++;
            result = true;
        }
        
        return result;

    }
    
    boolean comprar (Jugador jugador){
        
        boolean result = false;

        if (!tienePropietario()){
            propietario = jugador;
            result = true;
            jugador.paga(precioCompra);
        }

        return result;
        
    } 
    
    boolean vender(Jugador player){
        
        if (esEsteElPropietario(player) && !hipotecado){
            propietario.recibe(getPrecioVenta());
            derruirCasas(numCasas, propietario);
            numHoteles = 0;
            propietario = null;
            return true;
        }
        else
            return true;
    }
    
    public boolean esEsteElPropietario(Jugador player){        
        return propietario == player;        
    }
    
    void actualizaPropieatarioPorConversion(Jugador jugador){
        propietario = jugador;
    }
    
    public boolean getHipotecado(){
        
        return hipotecado;
    
    }
    
    private float getImporteHipoteca(){
        return (float) (hipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5)));
    }
    
    int getNumCasas(){
        return numCasas;
    }
    
    int getNumHoteles(){
        return numHoteles;
    }

    float getPrecioCompra(){
        return precioCompra;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    
    Jugador getPropietario(){
        return propietario;
    }
    
    boolean tienePropietario(){
        return propietario != null;
    }
    
   
}
